executor/join.go

```diff
@@ -180,16 +180,40 @@ func (e *HashJoinExec) getJoinKeyFromChkRow(isOuterKey bool, row chunk.Row, keyB
 
 	for _, i := range keyColIdx {
 		if row.IsNull(i) {
-			return true, keyBuf, nil
+			return true, false, keyBuf, nil
 		}
 	}
 
 	keyBuf = keyBuf[:0]
-	keyBuf, err = codec.HashChunkRow(e.ctx.GetSessionVars().StmtCtx, keyBuf, row, allTypes, keyColIdx)
-	if err != nil {
-		err = errors.Trace(err)
+	sc := e.ctx.GetSessionVars().StmtCtx
+	for idx, i := range keyColIdx {
+		d := row.GetDatum(i, allTypes[i])
+		if !isOuterKey {
+			outerType := e.outerExec.retTypes()[e.outerKeyColIdx[idx]]
+			// convert inner key to the type of outer key
+			converted, err := d.ConvertTo(sc, outerType)
+			if err != nil {
+				return false, false, keyBuf, errors.Trace(err)
+			}
+			cmp, err := converted.CompareDatum(sc, &d)
+			if err != nil {
+				return false, false, keyBuf, errors.Trace(err)
+			}
+			if cmp != 0 {
```

Comment:
Can you give an example where the converted datum is unequal with the original one?